File: Chapter4/buildOrder.js
/*

Build Order: You are given a list of projects and a list of dependencies (which
is a list of pairs of projects, where the second project is dependent on the
first project). All of a project's dependencies must be built before the project
is. Find a build order that will allow the projects to be built. If there is no
valid build order, return an error.

*/

class Node {
  constructor(val) {
    this.value;
    this.in1 = [];
    this.out = [];
    this.in = 0;
  }
}

class Graph {
  constructor(projects, dependencies) {
    this.projects = projects;
    this.dependencies = dependencies;
    this.nodes = {};

    this.createNodesHash();
    this.addDependencies();
  }

  addDependencies() {
    for (var i = 0; i < this.dependencies.length; i++) {
      this.nodes[this.dependencies[i][0]].out.push(this.dependencies[i][1]);
      this.nodes[this.dependencies[i][1]].in1.push(this.dependencies[i][0]);
      this.nodes[this.dependencies[i][1]].in += 1;
    }
  }

  createNodesHash() {
    for (let i = 0; i < this.projects.length; i++) {
      this.nodes[this.projects[i]] = new Node(this.projects[i]);
    }
  }
}

function buildOrder(projects, dependencies) {
  let projectHash = {};
  let result = [];

  for (let i = 0; i < projects.length; i++) {
    projectHash[projects[i]] = [];
  }

  // console.log(projectHash);
  for (let i = 0; i < dependencies.length; i++) {
    projectHash[dependencies[i][0]].push(dependencies[i][1]);
  }

  // console.log(projectHash);
  let notDone = true;
  while (notDone) {
    notDone = false;

    for (let i = 0; i < projects.length; i++) {
      if (projectHash[projects[i]].length == 0) {

      }
    }
  }
}

// console.log(buildOrder(["a", "b", "c", "d", "e", "f"], [["a", "d"], ["f", "b"], ["b", "d"], ["f", "a"], ["d", "c"]]));
let graph = new Graph(["a", "b", "c", "d", "e", "f"], [["a", "d"], ["f", "b"], ["b", "d"], ["f", "a"], ["d", "c"]]);
console.log(graph);

File: Chapter4/buildOrder.rb
# Build Order: You are given a list of projects and a list of dependencies
# (which is a list of pairs of projects, where the second project is dependent
# on the first project). All of a project's dependencies must be built before
# the project is. Find a build order that will allow the projects to be built.
# If there is no valid build order, return an error.

class Node
  attr_accessor :value, :out, :in

  def initialize(value)
    @value = value
    @out = []
    @in = 0
  end
end

class Graph
  attr_accessor :projects, :dependencies, :nodes

  def initialize(projects, dependencies)
    @projects = projects
    @dependencies = dependencies
    @nodes = {}

    createNodesHash()
    addDependencies()
  end

  def createNodesHash
    @projects.each do |el|
      @nodes[el] = Node.new(el);
    end
  end

  def addDependencies
    @dependencies.each do |el|
      @nodes[el[0]].out << el[1]
      @nodes[el[1]].in += 1
    end
  end
end

def buildOrder(projects, dependencies)
  graph = Graph.new(projects, dependencies)
  result = []
  len = projects.length

  while result.length < len
    invalid = true

    projects.each_with_index do |proj, idx|
      if graph.nodes[proj].in == 0
        curNodes = graph.nodes[proj].out

        curNodes.each do |curNode|
          graph.nodes[curNode].in -= 1
        end

        result << proj
        projects.delete_at(idx)
        invalid = false
        break
      end
    end

    return false if invalid
  end

  result
end

# graph = Graph.new(["a", "b", "c", "d", "e", "f"], [["a", "d"], ["f", "b"], ["b", "d"], ["f", "a"], ["d", "c"]])
# p graph
p buildOrder(["a", "b", "c", "d", "e", "f"], [["a", "d"], ["f", "b"], ["b", "d"], ["f", "a"], ["d", "c"]])
p buildOrder(["a", "b", "c", "d", "e", "f"], [["a", "d"], ["f", "b"], ["b", "d"], ["f", "a"], ["d", "c"], ["d", "a"]])
p buildOrder(["a", "b", "c", "d", "e", "f", "g"], [["f", "c"], ["f", "b"], ["f", "a"], ["c", "a"], ["b", "a"], ["b", "e"], ["a", "e"], ["d", "g"]])
